metabox/conftest.py

```diff
@@ -0,0 +1,181 @@
+    """
+    Custom pytest item for running a Scenario.
+    It loads the  appropriate machine configurations based on the scenario
+    variant and mode, sets up the machines, and runs the scenario.
+    It also handles cleanup and reporting of any failures that occur during the
+    run.
+    """
+    def __init__(self, *, scn, **kwargs):
+        super().__init__(**kwargs)
+        self.scn = scn
+
+    def __repr__(self) -> str:
+        return "<Scenario {}>".format(getattr(self, "name", None))
+
+    def _load(self, mode, release_alias):
+        config = self.config.runner.config[mode].copy()
+        config['alias'] = release_alias
+        config['role'] = mode
+        return self.config.runner.machine_provider.get_machine_by_config(
+            MachineConfig(mode, config))
+
+    def runtest(self):
+        scn = self.scn
+        if scn.mode == "remote":
+            scn.remote_machine = self._load("remote", scn.releases[0])
+            scn.service_machine = self._load("service", scn.releases[1])
+            scn.remote_machine.rollback_to('provisioned')
+            scn.service_machine.rollback_to('provisioned')
+            if scn.launcher:
+                scn.remote_machine.put(scn.LAUNCHER_PATH, scn.launcher)
+            scn.service_machine.start_user_session()
+        elif scn.mode == "local":
+            scn.local_machine = self._load("local", scn.releases[0])
+            scn.local_machine.rollback_to('provisioned')
+            if scn.launcher:
+                scn.local_machine.put(scn.LAUNCHER_PATH, scn.launcher)
+            scn.local_machine.start_user_session()
+        scn.run()
+        if not scn.has_passed():
+            if self.config.option.maxfail:
+                if scn.mode == "remote":
+                    msg = (
+                        "You may hop onto the target machines by issuing "
+                        "the following commands:\n{}\n{}").format(
+                            scn.remote_machine.get_connecting_cmd(),
+                            scn.service_machine.get_connecting_cmd())
+                elif scn.mode == "local":
+                    msg = (
+                        "You may hop onto the target machine by issuing "
+                        "the following command:\n{}").format(
+                            scn.local_machine.get_connecting_cmd())
+                print(msg)
+            else:
+                self.config.runner.machine_provider.cleanup()
+            raise MetaboxException(
+                self, self.scn.problems)
+        self.config.runner.machine_provider.cleanup()
+
+    def repr_failure(self, excinfo):
+        """Called when self.runtest() raises an exception."""
+        if isinstance(excinfo.value, MetaboxException):
+            err_msg = "Scenario execution failed\n    "
+            err_msg += "\n    ".join(excinfo.value.args[1])
+            return err_msg
+
+    def reportinfo(self):
+        return self.path, 0, f"Scenario: {self.name}"
```

Comment:
[The API](https://docs.pytest.org/en/latest/reference/reference.html#pytest.Item.reportinfo) mentions the default value for the line number is `None`, is there a rationale for using `0`?

Reply:
I borrowed this `conftest.py` from pytest [official documentation](https://docs.pytest.org/en/latest/example/nonpython.html). default is `None` but how does it renders? I'll check both